### ISSUE TITLE
TSX types missing

### ISSUE THREAD

**Issue by delaneyj** [thisbeyond/solid-dnd#68]
Ran into problems and looked at #60 but this example fails to run with typescript strict mode.

![image](https://user-images.githubusercontent.com/438252/202965795-4423e47c-ab59-4992-a129-61d3214698db.png)

With unhealthy amount of `any` I can compile
```ts
import {
  closestCenter,
  createSortable,
  DragDropProvider,
  DragDropSensors,
  DragOverlay,
  SortableProvider,
  useDragDropContext,
} from '@thisbeyond/solid-dnd'
import { Component, createSignal, For } from 'solid-js'

const Sortable: Component<{ item: any }> = (props) => {
  const sortable = createSortable(props.item)
  const ctx = useDragDropContext()
  return (
    <div
      use:sortable
      class="sortable"
      classList={{
        'opacity-25': sortable.isActiveDraggable,
        'transition-transform': !!ctx?.[0].active.draggable,
      }}
    >
      {props.item}
    </div>
  )
}

export const SortableVerticalListExample = () => {
  const [items, setItems] = createSignal([1, 2, 3])
  const [activeItem, setActiveItem] = createSignal(null)
  const ids = () => items()

  const onDragStart = (obj: any) => setActiveItem(obj.draggable.id)

  const onDragEnd = (obj: any) => {
    if (obj.draggable && obj.droppable) {
      const currentItems = ids()
      const fromIndex = currentItems.indexOf(obj.draggable.id)
      const toIndex = currentItems.indexOf(obj.droppable.id)
      if (fromIndex !== toIndex) {
        const updatedItems = currentItems.slice()
        updatedItems.splice(toIndex, 0, ...updatedItems.splice(fromIndex, 1))
        setItems(updatedItems)
      }
    }
  }

  return (
    <DragDropProvider
      onDragStart={onDragStart}
      onDragEnd={onDragEnd}
      collisionDetector={closestCenter}
    >
      <DragDropSensors />
      <div class="column self-stretch">
        <SortableProvider ids={ids()}>
          <For each={items()}>{(item) => <Sortable item={item} />}</For>
        </SortableProvider>
      </div>
      <DragOverlay>
        <div class="sortable">{activeItem()}</div>
      </DragOverlay>
    </DragDropProvider>
  )
}

```

However at runtime I'm getting ` useDragDropContext is not a function or its return value is not iterable` error and not sure what to do.  I know the lib is written in TS show I'd think the .tsx would be straight forward.  Could someone ELI5 what I'm missing?

**Comment by mohamed-adly** [comment]
Exact Same here

**Comment by mohamed-adly** [comment]
checkout #63

**Comment by jyasskin** [comment]
I filed the parallel issue https://github.com/thisbeyond/solid-dnd-site/issues/4 in the repository that holds the under-typed examples. It's not necessarily straightforward to fix this just because the library is in TS: giving users the types they expect is a whole extra development task compared to writing a library that's internally type-safe.

**Comment by ShivamJoker** [comment]
I am also unable to use it with TS, any possible fixes?

![image](https://user-images.githubusercontent.com/23727670/236510445-cf599e42-2733-4b13-b2b8-9af975638256.png)

**Comment by martinpengellyphillips** [comment]
Have you looked at #63 yet?

I'm general, you should be able to use the exported types from solid-dnd to get what you need (as well as the types from solidjs for things like components). 

Where specifically are the types not working for you (the screenshot you share is completely untyped)?

**Comment by Kapelianovych** [comment]
If you don't want to declare directives for **draggable**, **droppable** and **sortable** functions, you can pass them directly into the `ref` attribute, since directives are just a convenient way to obtain a reference to underlying node:

```jsx
<div ref={draggable}></div>
```

Of course, if you need the reference in another place too, the syntax becomes less plausible:

```jsx
let ref

<div ref={(element) => {
  ref = element;
  draggable(element);
}}></div>
```

But this should still work. Perhaps it is worth to add the definitions of those directives to the package itself, because it is a common feature that potentially is and will be used in most projects with DnD?

**Comment by ikemo3** [comment]
Here's how I fixed errors in strict mode: Specifically, the `useDragDropContext is not a function or its return value is not iterable` error was addressed by implementing `const state = context && context[0];`

```tsx
import {
  closestCenter,
  createSortable,
  DragDropProvider,
  DragDropSensors,
  DragEventHandler,
  DragOverlay,
  Id,
  SortableProvider,
  useDragDropContext,
} from "@thisbeyond/solid-dnd";
import { createSignal, For } from "solid-js";

const Sortable = (props: { item: Id }) => {
  const sortable = createSortable(props.item);
  const context = useDragDropContext();
  const state = context && context[0];  // <-- this

  return (
    <div
      ref={sortable}
      class="sortable"
      classList={{
        "opacity-25": sortable.isActiveDraggable,
        "transition-transform": !!state?.active.draggable,
      }}
    >
      {props.item}
    </div>
  );
};

export const SortableVerticalListExample = () => {
  const [items, setItems] = createSignal<Id[]>([1, 2, 3]);
  const [activeItem, setActiveItem] = createSignal<Id | null>(null);
  const ids = () => items();

  const onDragStart: DragEventHandler = ({ draggable }) =>
    setActiveItem(draggable.id);

  const onDragEnd: DragEventHandler = ({ draggable, droppable }) => {
    if (draggable && droppable) {
      const currentItems = ids();
      const fromIndex = currentItems.indexOf(draggable.id);
      const toIndex = currentItems.indexOf(droppable.id);
      if (fromIndex !== toIndex) {
        const updatedItems = currentItems.slice();
        updatedItems.splice(toIndex, 0, ...updatedItems.splice(fromIndex, 1));
        setItems(updatedItems);
      }
    }
  };

  return (
    <DragDropProvider
      onDragStart={onDragStart}
      onDragEnd={onDragEnd}
      collisionDetector={closestCenter}
    >
      <DragDropSensors />
      <div class="column self-stretch">
        <SortableProvider ids={ids()}>
          <For each={items()}>{(item) => <Sortable item={item} />}</For>
        </SortableProvider>
      </div>
      <DragOverlay>
        <div class="sortable">{activeItem()}</div>
      </DragOverlay>
    </DragDropProvider>
  );
};
```

**Comment by Koning3ton** [comment]
As done in the codebase of solid-dnd itself, you can use the '!' suffix to achieve the same:

```ts
const Sortable = (props: { item: Id }) => {
  const sortable = createSortable(props.item);
  const context = useDragDropContext()!; // <--

  //instead of
  const context = useDragDropContext();
  const state = context && context[0];  // <-- this
```
